### ISSUE TITLE
client stops playing connection reset

### ISSUE THREAD

**Issue by christf** [badaix/snapcast#436]
This is in a patched client and server using #432 and #433. The cause may be within those patches.

# Observations
* Clients stop playing without external interaction
*  in the server log I find:
```
2018-12-26 22-36-31 [Err] Exception in ControlSession::reader(): read_until: Connection reset by peer
```
* the client reports
```
2018-12-26 22-45-13 [Info] Chunk: 4     6       9       15      349     163
2018-12-26 22-45-14 [Info] Failed to get chunk
2018-12-26 22-45-18 [Notice] No chunk received for 5000ms. Closing ALSA.
```
* Restarting the client does not affect the symptom
* Restarting the *server* resolves the symptom, untouched clients start playing audio again
* When the problem exists, the server is still reading data from its pipe input 
* The symptom occurs with tcp_nodelay set or unset - tcp_nodelay does not affect the situation
* The encoder is still running according to debuglog
* according to netstat, the TCP connection is still ESTABLISHED with no data pending to be sent/read.
* with multiple clients the condition can occur for each client at a different time.
* in StreamSession::send() the following log appears when this happens. This means only messages o type 4 are sent.
```
2018-12-28 13-45-05 [Info] done: 2, size: 804, id: 0, refers: 0
2018-12-28 13-45-05 [Info] done: 2, size: 3133, id: 0, refers: 0
2018-12-28 13-45-05 [Info] done: 2, size: 225, id: 0, refers: 0
2018-12-28 13-45-05 [Info] done: 2, size: 226, id: 0, refers: 0
2018-12-28 13-45-05 [Info] done: 2, size: 982, id: 0, refers: 0
2018-12-28 13-45-07 [Info] done: 4, size: 8, id: 0, refers: 0
2018-12-28 13-45-11 [Info] done: 4, size: 8, id: 0, refers: 0
2018-12-28 13-45-12 [Info] done: 4, size: 8, id: 0, refers: 0
2018-12-28 13-45-16 [Info] done: 4, size: 8, id: 0, refers: 0
2018-12-28 13-45-17 [Info] done: 4, size: 8, id: 0, refers: 0
2018-12-28 13-45-21 [Info] done: 4, size: 8, id: 0, refers: 0
2018-12-28 13-45-22 [Info] done: 4, size: 8, id: 0, refers: 0
```

**Comment by christf** [comment]
Discarding data of a certain age in StreamSession::writer() is causing this. The current caching concept (playing data bufferMs milliseconds after the server has played it) does not consider that the client might be out of sync and thus be late, actually in need of old data.

The fix is in #433 

Unfortunately now with the fix (removal of the code that removes aged data) I am getting hissing after a while. This project is turning out to be more involved than I initially thought...


I am not sure yet what is causing the hiss. The clients report to be in sync. Both clients started playing what sounds like white noise simultaneously, so the cause must be in the server when generating the data.

**Comment by badaix** [comment]
fixed in v0.19.0
